ui/lib/core/addon/components/download-button.js

```diff
@@ -1,51 +1,67 @@
+ *   >
+ *    <Icon @name="download" />
+ *      Download
+ *   </DownloadButton>
+ * ```
+ * @param {string} data - data to download
+ * @param {boolean} [stringify=false] - argument to stringify the data before passing to the File constructor
+ * @param {string} [filename] - name of file that prefixes the ISO timestamp generated when download
+ * @param {string} [mime='text/plain'] - media type to be downloaded
+ * @param {string} [extension='txt'] - file extension
+ * @param {string} [buttonClass] - override css class defaults to customize how the button looks
+ */
 
-export default Component.extend({
-  layout: hbs`{{#if (has-block)}} {{yield}} {{else}} {{actionText}} {{/if}}`,
-  tagName: 'a',
-  role: 'button',
-  attributeBindings: ['role', 'download', 'href'],
-  download: computed('filename', 'extension', function () {
-    return `${this.filename}-${new Date().toISOString()}.${this.extension}`;
-  }),
+export default class DownloadButton extends Component {
+  get extension() {
+    return this.args.extension || 'txt';
+  }
 
-  fileLike: computed('data', 'mime', 'stringify', 'download', function () {
-    let file;
-    let data = this.data;
-    const filename = this.download;
-    const mime = this.mime;
-    if (this.stringify) {
-      data = JSON.stringify(data, null, 2);
-    }
-    if (window.navigator.msSaveOrOpenBlob) {
-      file = new Blob([data], { type: mime });
-      file.name = filename;
-    } else {
-      file = new File([data], filename, { type: mime });
-    }
-    return file;
-  }),
+  get mime() {
+    return this.args.mime || 'text/plain';
+  }
 
-  href: computed('fileLike', function () {
-    return window.URL.createObjectURL(this.fileLike);
-  }),
+  get filename() {
+    const defaultFilename = `${new Date().toISOString()}.${this.extension}`;
+    return this.args.filename ? this.args.filename + '-' + defaultFilename : defaultFilename;
+  }
 
-  click(event) {
-    if (!window.navigator.msSaveOrOpenBlob) {
-      return;
+  get data() {
+    if (this.args.stringify) {
+      return JSON.stringify(this.args.data, null, 2);
     }
-    event.preventDefault();
-    const file = this.fileLike;
-    //lol whyyyy
-    window.navigator.msSaveOrOpenBlob(file, file.name);
-  },
+    return this.args.data;
+  }
+
+  get content() {
+    const content = new File([this.data], this.filename, { type: this.mime });
+    return content;
+  }
 
-  actionText: 'Download',
-  data: null,
-  filename: null,
-  mime: 'text/plain',
-  extension: 'txt',
-  stringify: false,
-});
+  // TODO refactor and call service instead
+  @action
+  handleDownload() {
+    const { document, URL } = window;
+    const downloadElement = document.createElement('a');
+    downloadElement.download = this.filename;
+    downloadElement.href = URL.createObjectURL(this.content);
+    document.body.appendChild(downloadElement);
+    downloadElement.click();
+    URL.revokeObjectURL(downloadElement.href);
+    downloadElement.remove();
+  }
```

Comment:
Quite a bit of code was refactored in this component beyond glimmerizing but it looks more straightforward and I pulled down the branch and downloads work as expected so it looks good to me!

Reply:
Thank you for testing locally! 🎉